notebooks/ntsb-duckdb.ipynb

```diff
@@ -0,0 +1,340 @@
+    "df = df.rename(columns={'properties': 'metadata'})\n",
+    "in_memory_db.execute(f\"\"\"CREATE TABLE in_memory_table (doc_id {schema.get('doc_id')},\n",
+    "                      embeddings {schema.get('embeddings')}, metadata {schema.get('metadata')}, \n",
+    "                      text_representation {schema.get('text_representation')}, bbox {schema.get('bbox')}, \n",
+    "                      shingles {schema.get('shingles')}, type {schema.get('type')})\"\"\"\n",
+    "    )\n",
+    "in_memory_db.execute(\"\"\"INSERT INTO in_memory_table SELECT * FROM df; \n",
+    "                      CREATE INDEX in_memory_table_index ON in_memory_table USING HNSW(embeddings)\"\"\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# For queries, let's define an embedding function for the question that helps us easily compare the two vectors and generate optimal results // Run NN search\n",
+    "from sentence_transformers import SentenceTransformer\n",
+    "minilm = SentenceTransformer(\"sentence-transformers/all-MiniLM-L6-v2\")\n",
+    "\n",
+    "class Embedder():\n",
+    "    def __init__(self, llm):\n",
+    "        self.llm = llm\n",
+    "    \n",
+    "    def embed_query(self, question):\n",
+    "        v = self.llm.encode(question).tolist()\n",
+    "        return v\n",
+    "\n",
+    "embedder = Embedder(minilm)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Use DuckDB's ANN HNSW semantic search for retrieval. We use Array Euclidean Distance as the nearest neighbor lookup\n",
+    "q = \"What automobile type is the most accident prone?\"\n",
+    "in_memory_db.sql(f\"SELECT doc_id, text_representation, metadata FROM in_memory_table ORDER BY array_distance(embeddings, {embedder.embed_query(q)}::FLOAT[384]) LIMIT 10;\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Pretty print tables\n",
+    "from tabulate import tabulate\n",
+    "\n",
+    "q = \"What automobile type is the most accident prone?\"\n",
+    "result = in_memory_db.sql(f\"SELECT doc_id, text_representation, metadata FROM in_memory_table ORDER BY array_distance(embeddings, {embedder.embed_query(q)}::FLOAT[384]) LIMIT 10;\")\n",
+    "\n",
+    "# Convert the result to a list of lists\n",
+    "table_data = result.fetchall()\n",
+    "\n",
+    "# Get column names\n",
+    "headers = [desc[0] for desc in result.description]\n",
+    "\n",
+    "# Pretty print the table\n",
+    "print(tabulate(table_data, headers=headers, tablefmt=\"grid\"))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# View the first two rows of the text representation to visually confirm results \n",
```

Comment:
Move to end?

Reply:
Discussed on chat, left it there since it helps users intuitively make sense of the result